javascript/organizing_your_javascript_code/organizing_code_with_objects.md

```diff
@@ -0,0 +1,374 @@
+
+```javascript
+const myObject = {
+  property: 'Value!',
+  otherProperty: 77,
+  "obnoxious property": function() {
+    // do stuff!
+  }
+};
+```
+
+There are also 2 ways to get information out of an object: dot notation and bracket notation.
+
+```javascript
+// dot notation
+myObject.property; // 'Value!'
+
+// bracket notation
+myObject["obnoxious property"]; // [Function]
+```
+
+Which method you use will depend on context. Dot notation is cleaner and is usually preferred, but there are plenty of circumstances when it is not possible to use it. For example, `myObject."obnoxious property"` won't work because that property is a string with a space in it. Likewise, you cannot use variables in dot notation:
+
+```javascript
+const variable = 'property';
+
+myObject.variable; // this gives us 'undefined' because it's looking for a property named 'variable' in our object
+
+myObject[variable]; // this is equivalent to myObject['property'] and returns 'Value!'
+```
+
+If you are feeling rusty on using objects, now might be a good time to go back and review the content in our [object basics lesson](https://www.theodinproject.com/lessons/foundations-object-basics) from our JavaScript Basics course.
+
+### Lesson overview
+
+This section contains a general overview of topics that you will learn in this lesson.
+
+- Using objects to organize data.
+- Using objects to organize functionality.
+- Object methods and the `this` keyword
+- Public and private interfaces
+
+### Objects as a data structure
+
+You've already been introduced to the basic use of a JavaScript object- storing related information with key/value pairs. This is one of the simplest ways you can begin to organize your code! Take these examples from a 'tic tac toe' game:
+
+```javascript
+// example one
+const playerOneName = "tim";
+const playerTwoName = "jenn";
+const playerOneMarker = "X";
+const playerTwoMarker = "O";
+
+// example two
+const playerOne = {
+  name: "tim",
+  marker: "X"
+};
+
+const playerTwo = {
+  name: "jenn",
+  marker: "O"
+};
+```
+
+At first glance, the first doesn't seem so bad.. and it actually takes fewer lines to write than the example using objects, but the benefits of the second approach are huge! Grouping related data together into objects allows you to pass the data around easily. Let me demonstrate:
+
+```javascript
+function printName(player) {
+  console.log(player.name);
+}
+```
+
+This is something that you just could NOT do with the example one setup. Instead, every time you wanted to print a specific player's name, you would have to remember the correct variable name and then manually console.log it:
+
+```javascript
+console.log(playerOneName);
+console.log(playerTwoName);
+```
+
+Again, this isn't *that* bad... but what if you *don't know* which player's name you want to print?
+
+```javascript
+function gameOver(winningPlayer){
+  console.log("Congratulations!");
+  console.log(winningPlayer.name + " is the winner!");
+}
+```
```

Comment:
This and the previous suggestion were part of the original lesson. There was an attempt made to keep this part as-is for the most part. I'll reread in a bit and see what I think.

Reply:
Ah, I see. Since we're making these bigger changes, I think this can be addressed at the same time, as I'm not fully convinced by this specific example for the point being made.

Reply:
I did a bit of editing on this section, and want to hear your thoughts.

It reinforces the idea that objects may be more convenient than loads of separate arguments, like you suggested. I also provided a better example (IMO) of the `gameOver` function that truly shows how utilizing objects might be superior in such a situation.

I had to escape the \`\`\` symbols to prevent issues with formatting, sorry!
```
### Objects as a data structure

You've already been introduced to the basic use of a JavaScript object - storing related information *in one place* with key/value pairs. This is one of the simplest ways you can begin to organize your code! Take these examples from a 'tic tac toe' game:

\`\`\`javascript
// without objects
const playerOneName = "tim";
const playerTwoName = "jenn";
const playerOneMarker = "X";
const playerTwoMarker = "O";
let playerOneWins = 0;
let playerTwoWins = 0;

// with objects
const playerOne = {
  name: "tim",
  marker: "X",
  wins: 0
};

const playerTwo = {
  name: "jenn",
  marker: "O",
  wins: 0
};
\`\`\`

At first glance, the first doesn't seem so bad.. and it actually takes fewer lines to write than the example using objects, but the benefits of the second approach are huge!

Firstly, grouping related data together into objects allows you to pass *all* the data around easily. Calling a function such as `printPlayerInfo` below without objects would become increasingly cumbersome as the data you want to access increases:

\`\`\`javascript
// without objects
function printPlayerInfo(name, marker, wins) {
   console.log(`Name: ${name} | Marker: ${marker} | Wins: ${wins}`);
}

printPlayerInfo(playerOneName, playerOneMarker, playerOneWins)

// with objects
function printPlayerInfo(player) {
  console.log(`Name: ${player.name} | Marker: ${player.marker} | Wins: ${player.wins}`);
}

printPlayerInfo(playerOne)
\`\`\`

Secondly, grouping data into objects allows you update values in a cohesive way. Consider a `gameOver` function that increments the amount of wins for a player by one point when they win:

\`\`\`javascript
// without objects
function gameOver(winningPlayerName) {
  // figure out which player has the name passed in:
  if(winningPlayerName === playerOneName) {
    // increment the first player's score if the name matches the first player
    playerOneWins++
  } else if (winningPlayerName === playerTwoName) {
    // increment the second player's score if the name matches the first player
    playerTwoWins++
  }
}

gameOver("tim")

// with objects
function gameOver(winningPlayer) {
  winningPlayer.wins++
}

gameOver(playerOne)
\`\`\`

This example may be a little contrived, but notice how the example using an object doesn't need to know about any other players or variables, it can simply increment the score of the player object it's given! This is great, because it means you won't need to update the function if you add more players to your game - something that you just could NOT do without objects!

Or, what if we aren't making a 2 player game, but something more complicated such as an online shopping site with a large inventory? In that case, using objects to keep track of each particular item's name, price, description and other things is the only way to go. You will continue to use and see objects used in this way throughout the curriculum.
```

Reply:
Will get to this in due course. Github markdown allows quadruple backticks for code blocks containing triple backtick blocks, so you won't need to escape them.

Reply:
Small updates made above, just to let you know.